File: other/old-competitive-programming/LeetCode/115 - Distinct Subsequences/115. Distinct Subsequences.py
from functools import lru_cache


class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        @lru_cache(maxsize=None)
        def solve(i, j):
            if j == len(t):
                return 1
            if i == len(s):
                return 0
            ans = 0
            if s[i] == t[j]:
                ans += solve(i + 1, j + 1)
                ans += solve(i + 1, j)
            else:
                ans += solve(i + 1, j)
            return ans

        return solve(0, 0)

File: other/old-competitive-programming/LeetCode/115 - Distinct Subsequences/115.DistinctSubsequences.cpp
#include <iostream>
#include <vector>
using namespace std;

int solve(int i, int j, vector<vector<int> > &dp, string s, string t)
{
    if (j == t.length())
    {
        return 1;
    }
    if (i == s.length())
    {
        return 0;
    }
    if (dp[i][j] != -1)
    {
        return dp[i][j];
    }
    if (s[i] == t[j])
    {
        dp[i][j] = solve(i + 1, j + 1, dp, s, t) + solve(i + 1, j, dp, s, t);
    }
    else
    {
        dp[i][j] = solve(i + 1, j, dp, s, t);
    }

    return dp[i][j];
}
int numDistinct(string s, string t)
{
    vector<vector<int> > dp(s.length(), vector<int>(t.length(), -1));
    int res = solve(0, 0, dp, s, t);
    return res;
}
int main()
{
    cout << numDistinct("rabbbit", "rabbit");
    return 0;
}